KAFKA-2411; remove usage of blocking channel

ijuma opened the pull request.

ijuma committed:
Introduce `ChannelBuilders.create` and use it in `ClientUtils` and `SocketServer`

ijuma committed:
Use `Selector` instead of `BlockingChannel` for controlled shutdown

ijuma committed:
Use `Selector` instead of `BlockingChannel` in `ControllerChannelManager`

ijuma commented:
Tests passed locally.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -72,30 +77,37 @@ class ControllerChannelManager (private val controllerContext: ControllerContext
+      metrics,
+      time,
+      //FIXME Replace random suffix by something better
+      s"controller-${config.brokerId}-broker-${broker.id}-${scala.util.Random.nextInt(1000)}",
```
What is the right way to handle this? The random suffix is there because an exception will be thrown in `registerMetric` if a broker is removed and then added again. The `Selector` doesn't allow the caller to control that in its current form.

gwenshap replied:
Can to expend on why would the broker get removed and re-added from here?

ijuma replied:
`ControllerChannelManager.removeBroker` is called in `BrokerChangeListener.handleChildChange`, the actual line is:

`deadBrokerIds.foreach(controllerContext.controllerChannelManager.removeBroker(_))`

gwenshap replied:
So, it looks like you have a selector for each controller->broker connection?
I'd expect one Selector for the controller and open and close connections for brokers.
This way the selector string will include the broker ID for the controller, but not for the brokers themselves. And we'll probably want to maintain metricsPerConnection so we can track those separately.

Does that make sense?

ijuma replied:
Yes, I understand that one selector per controller would be better. I was trying to avoid a complete redesign of `ControllerChannelManager` though. It currently uses one `RequestSendThread` per broker with a `BlockingQueue` and `BlockingChannel`. `ControllerChannelManager` has a public `sendRequest` method that adds to the relevant queue. `RequestSendThread` takes from this queue. My understanding is that a given Selector should be used from just one thread and hence the resulting code. Please let me know if this is not correct.

In an ideal world, we would redesign things to use one Selector per Controller, but it seemed riskier and I thought it would be better to do that in a separate change after the next release (since our main goal at this point is just to use `Selector` for the TLS/SSL support). What are your thoughts?

gwenshap replied:
Lets see if I got it.

We have a communication thread per broker, and a selector per thread.
If a broker dies, we close the thread and the selector.
If the broker comes back, we can't create a new selector with the same name due to metrics.

Since the number of possible brokers is limited, can we just store the selector for each broker in a Map and reuse / reconnect if the dead broker is revived?  This should keep the metrics sane too.

ijuma replied:
Yes, that seems doable. I'll try that tomorrow and update the PR.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -72,30 +77,37 @@ class ControllerChannelManager (private val controllerContext: ControllerContext
-    val requestThread = new RequestSendThread(config.brokerId, controllerContext, broker, messageQueue, channel)
+    debug("Controller %d trying to connect to broker %d".format(config.brokerId, broker.id))
+    val selector = new Selector(
+      config.socketRequestMaxBytes,
```
Is it OK to use the same config as `SocketServer` for this?

gwenshap replied:
I don't think you need maxBytes here... this was for limiting the message size from clients. We have NetworkReceive.UNLIMITED if we want to leave this open.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -129,30 +143,32 @@ class RequestSendThread(val controllerId: Int,
-            receive = channel.receive()
+            selector.send(new NetworkSend(connectionId, RequestOrResponseSend.serialize(request)))
+            def findResponse = selector.completedReceives.asScala.find(_.source() == connectionId)
+            receive = SelectorUtils.pollUntilFound(selector, config.controllerSocketTimeoutMs)(findResponse)(time).getOrElse {
```
`config.controllerSocketTimeoutMs` was previously being used for `BlockingChannel.readTimeoutMs`, is it correct to use it here in this way?

gwenshap replied:
yeah, I think it does exactly the same thing we used to do: block until we get a reply or get to the timeout time.

IIRC, the NetworkReceive code has some deprecated methods specifically to support the blocking channel timeouts. Maybe this can be cleaned up now.

ijuma replied:
`BlockingChannel` is still used by the old consumer, so I am not sure we can remove anything just yet. I'll double-check though.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -165,20 +181,31 @@ class RequestSendThread(val controllerId: Int,
+        connectionId,
+        new InetSocketAddress(brokerEndPoint.host, brokerEndPoint.port),
+        config.socketSendBufferBytes,
+        config.socketReceiveBufferBytes
```
Is it OK to reuse the `SocketServer` settings here?

gwenshap replied:
BlockingChannel used to get the buffer sizes as parameters... shouldn't we keep the same behavior?

ijuma replied:
We were passing `BlockingChannel.UseDefaultBufferSize` (which is `-1`) everywhere and the code in `BlockingChannel` does:

```
        if(readBufferSize > 0)
          channel.socket.setReceiveBufferSize(readBufferSize)
        if(writeBufferSize > 0)
          channel.socket.setSendBufferSize(writeBufferSize)
```

So, we were basically not setting anything.

gwenshap replied:
LOL, we are cute :)

I'd leave it at -1 then.

ijuma replied:
We can't pass -1 to `Selector.connect` though as it will fail when it calls `Socket.sendBufferSize`. Maybe you are suggesting that I update `Selector.connect` to behave as `BlockingChannel` did when a negative value is passed in (basically not call `set*BufferSize`)?

gwenshap replied:
yep. Basically move the constants and their behavior to the new network classes.

ijuma replied:
Sounds good.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -165,20 +181,31 @@ class RequestSendThread(val controllerId: Int,
+        config.socketReceiveBufferBytes
+      )
+      val succeeded = SelectorUtils.pollUntil(selector,
+        config.controllerSocketTimeoutMs)(selector.connected().contains(connectionId))(time)
```
`config.controllerSocketTimeoutMs` was previously being used for `BlockingChannel.readTimeoutMs`, is it correct to use it here in this way?

gwenshap replied:
I think it is, since it keeps the same behavior.

ijuma commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -282,38 +304,46 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime) extends Logg
+                  connectionId(broker),
+                  new InetSocketAddress(brokerEndPoint.host, brokerEndPoint.port),
+                  config.socketSendBufferBytes,
+                  config.socketReceiveBufferBytes
```
Is it OK to use the `SocketServer` config parameters here?

gwenshap commented on core/src/main/scala/kafka/utils/SelectorUtils.scala:
```diff
@@ -0,0 +1,62 @@
+   * Invokes `selector.poll` until `find` returns `Some` or the timeout expires. It returns the result of `find` if
+   * the call completes normally or `None` if the timeout expires.
+   */
+  def pollUntilFound[T](selector: Selector, timeout: Long)(find: => Option[T])(implicit time: JTime): Option[T] = {
```
I think these utils are very generic, but also make the code a bit harder to read.

It looks like the "find" we are waiting for is basically always a reply from a specific broker? In this case, maybe make the generic method private and provide a more specific wrapper?

gwenshap replied:
I'd also document that these are for creating blocking behavior on top of our async network classes, and to use only where blocking makes sense...

ijuma replied:
I can try to add some wrappers and see how it looks. To help me understand, what makes it hard to read? Is it because of the long line with many parameters? This could be simplified quite a bit by using implicit value classes and passing the time implicitly as well. Example follows:

Now:

```
receive = SelectorUtils.pollUntilFound(selector, config.controllerSocketTimeoutMs)(findResponse)(time).getOrElse {
```

After the proposed change:

```
receive = selector.pollUntilFound(config.controllerSocketTimeoutMs)(findResponse).getOrElse {
```

ijuma replied:
Good point about documenting the blocking aspect. Will do.

gwenshap replied:
I actually find the very generic "any predicate" or "any find on a collection" challenging. I'm concerned that future contributors won't quite know what to do with those.

I'm thinking that baking the specific "find" we actually have into a wrapper, will help.
So:
pollUntilRecievingResponseFromConnection(selector, timeout, connection_id)

ijuma committed:
Move `readCompletely` from `NetworkReceive` to `BlockingChannel`

It is now a private method since it's not used anywhere else and it's
been changed slightly to match the use-case better.

ijuma committed:
Adjust buffer and max request size to match `BlockingChannel` behaviour

Based on feedback from Gwen.

ijuma committed:
Introduce specific methods in `SelectorUtils` and make the generic ones private

As suggested by Gwen.

ijuma committed:
Reuse `Selector` when removing and re-adding brokers in `ControllerChannelManager`

As suggested by Gwen.

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2411-remove-usage-of-blocking-channel

* apache/trunk:
  KAFKA-2330: Vagrantfile sets global configs instead of per-provider override configs; patched by Ewen Cheslack-Postava, reviewed by Geoff Anderson and Gwen Shapira
  KAFKA-2246; Fix incorrect config ZK path.
  KAFKA-2084; trivial follow-up (remove JUnit3Suite dependency)

ijuma commented:
@gwenshap, I believe I have addressed everything that was discussed. I also merged trunk into this branch to fix a minor import conflict in a test. Tests pass locally. Please have a look and if you're happy, this may be good to go.

ijuma commented:
Closed/reopened to trigger another CI build.

ijuma commented:
The output doesn't actually say what (if anything) failed in the last CI build.

ijuma commented:
Ran the tests locally two more times and they passed both times.

gwenshap commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -44,7 +49,7 @@ class ControllerChannelManager (private val controllerContext: ControllerContext
   def shutdown() = {
     brokerLock synchronized {
-      brokerStateInfo.foreach(brokerState => removeExistingBroker(brokerState._1))
+      brokerStateInfo.values.foreach(removeExistingBroker)
```
nice :)

gwenshap commented:
I think this needs a rebase, but otherwise LGTM. @junrao, do you want to take a look too?

gwenshap commented:
actually, not good - maybe rebase related - but cloning the PR doesn't pass unit tests on my machine:
uncaught exception during compilation: java.lang.StackOverflowError

(Scala 2.10.5, Java 8)

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2411-remove-usage-of-blocking-channel

* apache/trunk:
  KAFKA-1901; Move Kafka version to be generated in code by build (instead of in manifest); reviewed by Ismael Juma, Joel Koshy, Jason Rosenberg
  KAFKA-2015: Enable ConsoleConsumer to use new consumer

gwenshap commented:
ok, the stack size issue is intermittent and seem unrelated to the patch (happened to me once or twice on trunk too). My LGTM is back. @junrao wanted to take a look, so I'll hold off committing to give him a chance to comment.

ijuma commented:
I fixed the conflict (just an import)

gwenshap commented:
Cool. The test failure seems like an unrelated flaky. Opened KAFKA-2455 for it.

ijuma commented:
Thanks!

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -283,38 +305,44 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime) extends Logg
+                  new InetSocketAddress(brokerEndPoint.host, brokerEndPoint.port),
+                  Selectable.USE_DEFAULT_BUFFER_SIZE,
+                  Selectable.USE_DEFAULT_BUFFER_SIZE
+                )
```
We need to block until the connection is established, right?

ijuma replied:
I thought we did not as we will eventually block after the `send`, which happens a few lines below. Have I misunderstood how the `Selector` should be used in this case?

junrao replied:
If you look at NetworkClient, the way that it uses Selector is the following. Before NetworkClient can send any request, it has to make sure the connector is connected and the channel is ready. If you don't follow this protocol, things can get weird. For example, in SSL, after we finish the handshake, we will turn off the interest bit for write. If the interest bit for write has already been turned on by a request sent before the handshake completes, the send may never complete.

Thinking a bit more about this. It seems to implement a blocking channel on the Selector, we will need most of the connection state management logic in NetworkClient. The problem with using NetworkClient is that it's tied to Metadata refresh, which is not needed. We can either add an option to turn off the metadata part in NetworkClient and use it to implement the blocking channel. Alternatively, we can implement a BlockingSelector on top of Selector, but has to copy some of the state management logic from NetworkClient over. Not sure which one is better.

ijuma replied:
OK, I see. I'll check if turning off metadata in `NetworkClient` can be done without too much complexity.

junrao commented on core/src/main/scala/kafka/utils/SelectorUtils.scala:
```diff
@@ -0,0 +1,89 @@
+    if (timeout < 0) {
+      do {
+        selector.poll(timeout)
+        result = find
```
We will need to handle disconnects (like what NetworkClient does) by checking selector.disconnected(). If a socket is disconnected, there is no need to wait for the timeout. We should just throw an IOException back. Ditto to the selector.poll() below.

ijuma replied:
Makes sense, will do.

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -72,30 +77,38 @@ class ControllerChannelManager (private val controllerContext: ControllerContext
+        config.connectionsMaxIdleMs,
+        metrics,
+        time,
+        s"controller-${config.brokerId}-broker-${broker.id}",
```
We haven't been using the s notation so far. To be consistent, perhaps it's better to use the string format thing for now and do a global replacement at some point if we feel that's better?

ijuma replied:
I considered that, but we have over 800 instances of `format` in our codebase, so it would be quite a painful global replacement (with potential for introducing bugs). String interpolation is safer and performs better than `.format`, so I thought we could use it for new code where it made sense. We use it in a small number of places already. Please let me know if you still think I should change it.

ijuma replied:
@junrao, I added the `broker-id` tag below. Does it still make sense to include `broker.id` in the `metricGrpPrefix` above?

junrao replied:
metricGrpPrefix is just a prefix of the metrics group, which is a string. Tags give metadata in key/value pairs, which is more informative.

junrao replied:
Now that we have the broker tag, the group prefix can just be "controller-channel" or sth like that.

ijuma replied:
OK. We don't need to include the controller id in the group prefix or as a tag?

junrao replied:
We don't need the broker id in group prefix. Still need the broker id in the metrics tag.

ijuma replied:
Sorry, my question was about the controller id (I had understood the point about the broker id).

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -72,30 +77,38 @@ class ControllerChannelManager (private val controllerContext: ControllerContext
+        metrics,
+        time,
+        s"controller-${config.brokerId}-broker-${broker.id}",
+        Map.empty.asJava,
```
Since there is a selector per broker, we need to add a broker id tag for the metric tag.

ijuma replied:
Will do. I'm not really sure how these tags work to be honest, but I'll see if I can find an existing example. :)

ijuma committed:
Handle disconnections in `SelectorUtils`

As suggested by Jun.

ijuma committed:
Add broker id tag to `metricTags` passed to `Selector`

As suggested by Jun

ijuma commented:
@junrao, thanks for the review. Learned a couple of new things about the `Selector` and metrics. :) I pushed two commits to address two of your comments. The other two I was not sure about so I asked some questions.

ijuma commented:
Closing this for now.

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2411-remove-usage-of-blocking-channel

* apache/trunk:
  KAFKA-2439; Add MirrorMaker service class for system tests
  KAFKA-2457; Fix how the argument is passed to `compileScala`
  KAFKA-2457; StackOverflowError during builds

ijuma committed:
Make metadata updates optional in `NetworkClient`

ijuma committed:
Add support for `ControlledShutdown` in `Protocol`

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into wip/kafka-2411-remove-usage-of-blocking-channel

* apache/trunk:
  KAFKA-2467: Fix changes to behavior in ConsoleConsumer: properly parse consumer.config option, handle exceptions during message processing, and print number of processed messages to stderr.
  KAFKA-2388: refactor KafkaConsumer subscribe API
  KAFKA-2072: Add StopReplica request/response to o.a.k.common.requests
  KAFKA-1683; persisting session information in Requests
  KAFKA-2136; Add throttle time (on quota violation) in fetch/produce responses; reviewed by Joel Koshy, Dong Lin and Jun Rao
  KAFKA-2351; Catch all exceptions in socket server's acceptor; reviewed by Grant Henke, Joel Koshy, Jiangjie Qin, Jun Rao

ijuma committed:
Use `NetworkClient` instead of `Selector` for controlled shutdown

ijuma committed:
Add support for `UpdateMetadata` and `LeaderAndIsr` in `Protocol`

ijuma committed:
Use `NetworkClient` instead of `Selector` for controller/broker communication

ijuma commented:
Tests passed locally.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/ManualMetadataUpdater.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ManualMetadataUpdater implements MetadataUpdater {
```
Will add javadoc.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/MetadataUpdater.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.List;
+
+interface MetadataUpdater {
```
Will add javadoc.

junrao replied:
Is that added?

ijuma replied:
Not yet, I was focusing on getting the behaviour right first. I'll add all the missing javadoc/scaladoc in my next commit.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -304,6 +329,18 @@ public RequestHeader nextRequestHeader(ApiKeys key) {
     }
 
     /**
+     * Generate a request header for the given API key
```
Will add `and version` at the end of the sentence.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -465,7 +517,58 @@
+
+    public static final Schema UPDATE_METADATA_REQUEST_END_POINT_V1 =
+            // for some reason, V1 sends `port` before `host` while V0 sends `host` before `port
+            new Schema(new Field("port", INT32, "The port on which the broker accepts requests."),
```
V1 was added during the development of 0.8.3 so we could change this to be like V0 if we think it's important.

junrao replied:
Good catch. It's probably too late to change that though.

ijuma commented on core/src/main/scala/kafka/utils/NetworkClientBlockingOps.scala:
```diff
@@ -0,0 +1,102 @@
+    new NetworkClientBlockingOps(client)
+}
+
+class NetworkClientBlockingOps(val client: NetworkClient) extends AnyVal {
```
Will add `scaladoc` to this class and methods.

junrao replied:
We probably want to move the blocking support to the client side. In the future, we likely will be writing the admin tools in java and it potentially will need to make blocking request calls too.

ijuma replied:
Is this something we want to do now or when the need arises? I didn't want to add this to the clients jar because I thought it was just useful to help us transition the broker code without bigger changes. It seems like there may be more use cases, but isn't it better to wait until they are concrete before moving the code to the clients jar?

junrao replied:
We can keep the blocking logic in scala for now.

ijuma commented:
Tests passed locally and in Jenkins once. The second time failed due to an unrelated failure that has been seen in other PRs (testMetricsLeak).

junrao commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -81,6 +81,14 @@
 
     /**
+     * Disconnect connection to a particular node (if there is one). The disconnection is asynchronous and will not be
+     * processed until the next {@link #poll(long, long) poll()} call.
```
Do we need to wait for the next poll() call? disconnect() will cancel the key, after which the key will never be selected again.

ijuma replied:
I copied that documentation from `Selector.disconnect`. There is the following code in `Selector.poll`:

```
                    if (!key.isValid()) {
                        close(channel);
                        this.disconnected.add(channel.id());
                    }
```

And there is the following test in `NetworkClientTest`:

```
        assertTrue("Now the client is ready", client.ready(node, time.milliseconds()));
        selector.disconnect(node.idString());
        client.poll(1, time.milliseconds());
        selector.clear();
        assertFalse("After we forced the disconnection the client is no longer ready.", client.ready(node, time.milliseconds()));
        assertTrue("Metadata should get updated.", metadata.timeToNextUpdate(time.milliseconds()) == 0);
```

So, that lead me to believe that a poll was needed to update the internal state of the `Selector` and `NetworkClient`. The test above fails without the `poll` line (although it does use a `MockSelector`, so the bug could be there).

What are your thoughts based on this additional information?

junrao replied:
Yes, that's how MockSelector is implemented. In the real selector, my understanding is that if you disconnect a client, we cancel the key. The cancelled key will never be selected by the selector again. So, you would have to set the connection state in the same disconnect call. If that's the case, we probably need to change the behavior of MockSelector.

ijuma replied:
I see. I'll test the behaviour of the real selector just to be sure. In the likely case that it matches your description, I will do the following:
- Update `MockSelector`
- Update `NetworkClientTest`
- Update `Selector.disconnect` and `NetworkClient.disconnect` documentation
- Update implementation of `NetworkClient.disconnect` to update connection states
- Remove `NetworkClientBlockingOps.disconnectAndPoll` as `disconnect` will be sufficient

Have I covered everything?

junrao replied:
That sounds good.

junrao commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -119,6 +145,17 @@ public boolean ready(Node node, long now) {
+     */
+    @Override
+    public void disconnect(String nodeId) {
+        selector.disconnect(nodeId);
```
It seems that we also need to remove nodeId from ClusterConnectionStates?

ijuma replied:
I did think about setting the state to DISCONNECTED (which has a similar effect), but then I thought that this would be updated automatically via `poll` as per the comment added in the discussion around the javadoc for `KafkaClient.poll`. But maybe the right thing is to do update the state here as you suggest. Do you think we should remove the id or set the state to `DISCONNECTED`?

junrao replied:
This is related to the comment above. If a client calls disconnect, it may not be interested in using this connection any more. So, we probably want to fresh the memory by removing the id from the connection state.

junrao commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -510,9 +476,143 @@ private void initiateConnect(Node node, long now) {
+            return metadata.fetch().nodes();
+        }
+
+        public boolean isUpdateDue(long now) {
```
Need to add @override.

ijuma replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -510,9 +476,143 @@ private void initiateConnect(Node node, long now) {
+            return !this.metadataFetchInProgress && this.metadata.timeToNextUpdate(now) == 0;
+        }
+
+        public long maybeUpdate(long now) {
```
Need to add @override.

ijuma replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -510,9 +476,143 @@ private void initiateConnect(Node node, long now) {
+            return false;
+        }
+
+        public void requestUpdate() {
```
Need to add @override.

ijuma replied:
Will fix.

junrao commented on clients/src/main/java/org/apache/kafka/common/network/ChannelBuilders.java:
```diff
@@ -0,0 +1,52 @@
+     * @return the configured `ChannelBuilder`
+     * @throws IllegalArgumentException if `mode` invariants described above is not maintained
+     */
+    public static ChannelBuilder create(SecurityProtocol securityProtocol, SSLFactory.Mode mode, Map<String, ?> configs) {
```
Do we need to pass mode in? Could we just get it from configs?

ijuma replied:
The 4 places where this is used all pass this parameter explicitly (as `SSLFactory.Mode.{Client,Server}`). We could put a key into the configs map, but I think that would make things more opaque with no clear benefit. What do you think?

junrao replied:
What you have is fine. I didn't realize that mode is implicit and not an explicit property.

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -442,6 +461,39 @@
+                                                                      new Field("controller_epoch", INT32, "The controller epoch."),
+                                                                      new Field("partition_states",
+                                                                                new ArrayOf(LEADER_AND_ISR_REQUEST_PARTITION_STATE_V0)),
+                                                                      new Field("leaders", new ArrayOf(LEADER_AND_ISR_REQUEST_LEADER_V0)));
```
This field should be named live_leaders.

ijuma replied:
Will fix. We also have `alive_brokers` elsewhere, should we be calling that `live_brokers` instead for consistency?

junrao commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -465,7 +517,58 @@
+    /* Update metadata api */
+
+    public static final Schema UPDATE_METADATA_REQUEST_PARTITION_STATE_V0 =
+            new Schema(new Field("topic", STRING, "Topic name."),
```
This can just be referencing LEADER_AND_ISR_REQUEST_PARTITION_STATE_V0.

ijuma replied:
OK, will do. I wasn't sure what was our policy when it came to referencing schemas from other APIs in the same class. Good to know.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrRequest.java:
```diff
@@ -0,0 +1,212 @@
+        return partitionStates;
+    }
+
+    public Set<EndPoint> leaders() {
```
leaders is better named as liveLeaders.

ijuma replied:
Will do.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java:
```diff
@@ -0,0 +1,250 @@
+    private final Set<Broker> aliveBrokers;
+
+
+    public UpdateMetadataRequest(int version, int controllerId, int controllerEpoch, Map<TopicPartition,
```
It's probably better to create two constructors, one for each version. We can then mark the v0 constructor as deprecated and can remove it in the future.

ijuma replied:
Will do.

ijuma replied:
I did this, but it's worth being aware of the downsides (and why I had avoided that in the first place):
- We now have a number of warnings due to the fact that we call the deprecated constructor
- The calling code is more complex as it needs to pass slightly different data to each constructor (we still support both versions from the calling code and previously it just had to pass a version)
- A new deprecated inner class `BrokerEndPoint` had to be introduced for the V0 constructor

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java:
```diff
@@ -0,0 +1,250 @@
+            int brokerId = brokerData.getInt(BROKER_ID_KEY_NAME);
+
+            // V0
+            if (brokerData.hasField(HOST_KEY_NAME)) {
```
Both v0 and v1have HOST_KEY_NAME. Perhaps we should check field SECURITY_PROTOCOL_TYPE_KEY_NAME?

ijuma replied:
In V1, `HOST_KEY_NAME` is in the `end_points` struct though, so it should be fine either way as far as I can see. I'll change it to `SECURITY_PROTOCOL_TYPE_KEY_NAME` as it's clearer though.

ijuma replied:
Actually I can't make this change because `SECURITY_PROTOCOL_TYPE_KEY_NAME` will never exist at this level. I could check for `ENDPOINTS_KEY_NAME`, but I think the current way is less error-prone (as endpoints could be empty due to a bug perhaps). So I'll leave it as is unless you disagree.

junrao replied:
What you had is fine.

junrao commented on clients/src/test/java/org/apache/kafka/clients/MockClient.java:
```diff
@@ -201,6 +201,11 @@ public RequestHeader nextRequestHeader(ApiKeys key) {
     }
 
     @Override
+    public RequestHeader nextRequestHeader(ApiKeys key, short version) {
```
Should we mock the other new method disconnect() too?

ijuma replied:
It was already there previously.

junrao commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -70,18 +75,37 @@ public void testSerialization() throws Exception {
-                createStopReplicaResponse());
+                createStopReplicaResponse(),
+                createUpdateMetadataRequest(1),
+                createUpdateMetadataRequest(1).getErrorResponse(0, new UnknownServerException()),
```
response version should be 1.

ijuma replied:
Good catch, will fix.

junrao commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -70,18 +75,37 @@ public void testSerialization() throws Exception {
+                createUpdateMetadataRequest(1).getErrorResponse(0, new UnknownServerException()),
+                createUpdateMetadataResponse(),
+                createLeaderAndIsrRequest(),
+                createLeaderAndIsrRequest().getErrorResponse(0, new UnknownServerException()),
```
response version should be 1.

ijuma replied:
Good catch, will fix.

ijuma replied:
I checked this and I think 0 is correct for `LeaderAndIsrResponse`. Or am I missing something?

ijuma replied:
I think you meant `ControlledShutdownResponse`, I fixed that.

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -72,30 +79,48 @@ class ControllerChannelManager (private val controllerContext: ControllerContext
+        selector,
+        new ManualMetadataUpdater(Seq(brokerNode).asJava),
+        config.brokerId.toString,
+        Int.MaxValue,
```
We probably want maxInflightRequests to be 1 so that it's consistent with the current behavior in BlockingChannel.

ijuma replied:
I think the behaviour should be consistent anyway as we block for each request. Setting this to `1` will enforce it, so I'll change it.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -32,6 +37,8 @@ import kafka.utils.ZkUtils._
 import kafka.utils._
 import kafka.utils.CoreUtils._
 import org.apache.zookeeper.Watcher.Event.KeeperState
+import org.apache.kafka.common.metrics.Metrics
```
There are a few unused imports.

ijuma replied:
Will fix.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -47,12 +54,24 @@ class ControllerContext(val zkClient: ZkClient,
+  val partitionsUndergoingPreferredReplicaElection: mutable.Set[TopicAndPartition] = new mutable.HashSet
+
+  /**
+   * This map is used to ensure the following invariant: at most one `Selector` instance should be created per broker
```
selector should now be NetworkClient. There are a few other mentions of selector below.

ijuma replied:
I'll change this to say `NetworkClient`/`Selector`. I think the other `Selector` mentions are fine.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -262,17 +268,126 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime) extends Logg
+    val socketTimeoutMs = config.controllerSocketTimeoutMs
+
+    def socketTimeoutException: Throwable =
+      new SocketTimeoutException(s"Did not receive response within $socketTimeoutException")
```
$socketTimeoutException should be $socketTimeoutMs, right?

ijuma replied:
Good catch, will fix.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -262,17 +268,126 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime) extends Logg
+          selector,
+          metadataUpdater,
+          config.brokerId.toString,
+          Integer.MAX_VALUE,
```
We probably want maxInflightRequests to be 1 so that it's consistent with the current behavior in BlockingChannel.

ijuma replied:
Similar to the other instance of this, I think the behaviour is still correct, but I'll change it to be clearer/safer.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -262,17 +268,126 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime) extends Logg
+                  networkClient.disconnectAndPoll(node(prevController).idString, 0)
+
+                prevController = broker
+                metadataUpdater.setNodes(Seq(node(prevController)).asJava)
```
Do we need to do this? Nodes are only used for sending MetadataRequest and we are not doing that here.

ijuma replied:
We don't strictly need it, but `KafkaClient.leastLoadedNode` would return `null` if we don't do this. As you say, we don't use that method so it would have no effect right now if we remove it. I did that because I thought it may help avoid surprises in the future, but I don't have a strong opinion on it. Do you still think it should be removed?

junrao replied:
It's fine if we set the nodes explicitly. If so, we should do the same for the ManualMetadataUpdater used in ControllerChannelManager.

ijuma replied:
We do indeed, but at construction time because we use a `NetworkClient` for each broker connection, I paste the line below:

`new ManualMetadataUpdater(Seq(brokerNode).asJava)`

junrao replied:
Got it. I missed that you can pass in the nodes from the constructor.

junrao commented on core/src/main/scala/kafka/utils/NetworkClientBlockingOps.scala:
```diff
@@ -0,0 +1,102 @@
+      else if (client.connectionFailed(node))
+        throw new IOException(s"Connection to $node failed")
+      else {
+        retryBackoff.foreach(Thread.sleep)
```
If the connection didn't fail, is there any reason to back off? It seems that we should be calling networkClient.poll() with the timeout immediately.

ijuma replied:
I'll remove `retryBackoff`. It's how I had it at first, but I was seeing some issues where `poll` was returning immediately which was causing the thread to spin and seemed to be causing more connection timeouts than usual. I now think that this was caused by a bug in the handling of the connection failures (which I fixed before reopening the PR).

junrao commented on core/src/main/scala/kafka/utils/NetworkClientBlockingOps.scala:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  def blockingSendAndReceive(request: ClientRequest, timeout: Long)(implicit time: JTime): Option[ClientResponse] = {
+    client.send(request)
```
Perhaps it's more convenient to include to logic to wait for the connection being ready here. This way, we are guaranteed that every time we send a request, the socket is ready. Currently, in ControllerChannelManager, if we can't establish a socket connection, it seems that we can send a request before the channel is ready.

ijuma replied:
Perhaps. I'm not sure though. It's easy enough for the caller to call `blockingReady` if that is desired, right? I didn't do that in `ControllerChannelManager` because the existing code already handles the exception that would be thrown if the initial connection fails:

```
case e: Throwable => // if the send was not successful, reconnect to broker and resend the message
  warn(("Controller %d epoch %d fails to send request %s to broker %s. " +
    "Reconnecting to broker.").format(controllerId, controllerContext.epoch,
      request.toString, toBroker.toString()), e)
  networkClient.disconnectAndPoll(brokerNode.idString, 0)(time)
  connectToBroker()
  isSendSuccessful = false
  // backoff before retrying the connection and send
  CoreUtils.swallowTrace(Thread.sleep(300))
```

The existing code is not the prettiest, but I tried not to make too many changes at this point. I think we should consider redesigning the controller/broker communication to make better use of `NetworkClient` in the future.

junrao replied:
Yes, the logic in ControllerChannelManager is not pretty. The issue that I see is that if connectToBroker() fails to establish a connection in time, we just disconnect. So, by the time you send a request, there is not guarantee that the connection is ready.

ijuma replied:
Yes, I understand. So, if the connection is not ready, an `IllegalStateException` will be thrown which will cause a disconnect (no-op) followed by another connect and then we will retry again after a 300ms sleep since `isSendSuccessful === false`. This behaviour is very similar to what would happen with `BlockingChannel` where a ClosedChannelException would be thrown if `send` was called and the channel was not connected due to a failure.

I can add a `blockingReady` call before the `send` if you think that is better though.

junrao replied:
Yes, relying on IllegalStateException probably works, but it's kind of hacky. IllegalStateException is meant to capture all programming errors, and using it for error handling feels wrong.

ijuma replied:
OK, I will add a `blockingReady` before the send then.

ijuma committed:
Add missing @Override, remove unused imports and minor comment tweak

Many of these suggested by Jun.

ijuma committed:
Fixes and improvements to `UpdateMetadata`, `LeaderAndIsr` and `ControlledShutdown`

As suggested by Jun

ijuma committed:
Remove `retryBackoff` from `blockingReady`, reduce `maxInFlightRequestsPerConnection` to 1 and fix message in `SocketTimeoutException`

As suggested by Jun

ijuma commented:
Thanks for the review @junrao. I have addressed a number of the issues raised and left comments/questions for the rest.

ijuma commented:
`./gradlew test` passed locally (I ran it 3 times). However, there is one system test failure:

test_id:    2015-08-31--001.kafkatest.tests.replication_test.ReplicationTest.test_hard_bounce
status:     FAIL
run time:   7 minutes 32.756 seconds

http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/60/console

I will investigate this tomorrow (pointers are welcome though).

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2411-remove-usage-of-blocking-channel

* apache/trunk:
  KAFKA-2486; fix performance regression in new consumer
  KAFKA-2485; Allow producer performance to take properties from a file…
  KAFKA-2475: Make Copycat only have a Converter class instead of Serializer, Deserializer, and Converter.
  MINOR: kafkatest add manifest
  KAFKA-2468: SIGINT during Kafka server startup can leave server deadlocked
  KAFKA-2377: Add basic system test for copycat using source and sink file connectors.
  KAFKA-2367; Add Copycat runtime data API.

ijuma committed:
Check if broker is ready before sending and don't connect to broker in constructor

There are 2 slight changes in behaviour:

1. We no longer try to connect in the constructor of `RequestSendThread`
This fixes a potential thread-safety issue and it means that the connection
happens when the thread is started from ControllerChannelManager.startup.

2. If there is a failure, we backoff before we try to connect again
This seems slightly better than the previous behaviour of attempting
reconnection immediately and doing the backoff before the send.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java:
```diff
@@ -0,0 +1,291 @@
+    private final int controllerId;
+    private final int controllerEpoch;
+    private final Map<TopicPartition, PartitionState> partitionStates;
+    private final Set<Broker> aliveBrokers;
```
We have `liveLeaders` in `LeaderAndIsrRequest`. Do we want to keep this as `aliveBrokers` or would it be better to call it `liveBorkers`?

junrao replied:
Even though UpdateMetadataRequest_v0 has identical structure as LeaderAndIsrRequest_v0, the set of brokers used are slightly different. In UpdateMetadataRequest_v0, we pass in all live brokers. In LeaderAndIsrRequest_v0, we pass in all live brokers that are the leaders. So, we can keep the names as they are.

ijuma replied:
Yes, I understand the difference between brokers and leaders. The bit I was asking about is the `live` prefix (in `liveLeaders`) versus the `alive` prefix (in `aliveBrokers`). It seemed to me that `liveBrokers` means the same and the prefix would be consistent. Anyway, not a big deal, so fine to keep as is if you prefer.

junrao replied:
Yes, I agree that it's better to make the names consistent. So we can use live_brokers and live_leaders here and in protocols as well.

ijuma committed:
Replace `NetworkClient.disconnect(id)` with `NetworkClient.close(id)`

This works similarly to `Selector.close(id)`.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -129,30 +162,35 @@ class RequestSendThread(val controllerId: Int,
-            channel.send(request)
-            receive = channel.receive()
-            isSendSuccessful = true
+            if (!brokerReady()) {
```
I did this here instead of changing `blockingSendAndReceive` to make it easier to maintain the previous behaviour of logging once a connection is established.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -81,6 +81,13 @@
+     *
+     * @param nodeId The id of the node
+     */
+    public void close(String nodeId);
```
This method already existed in `NetworkClient`, I just added it to the interface.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -119,6 +145,17 @@ public boolean ready(Node node, long now) {
+     * @param nodeId The id of the node
+     */
+    @Override
+    public void close(String nodeId) {
```
I named this `close` instead of `disconnect` to match the relevant method in `Selector`, but I'd be interested in feedback regarding the naming. Also, it's not clear to me when `Selector.disconnect` is actually useful.

ijuma commented:
@junrao, pushed two additional commits to address the two outstanding issues. I implemented things in a slightly different way than what was discussed. I added some comments to the PR explaining. Let me know what you think.

`gradlew test` passed locally and system tests passed too:

http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/63/console

junrao commented on clients/src/main/java/org/apache/kafka/common/errors/BrokerNotAvailableException.java:
```diff
@@ -0,0 +1,32 @@
+
+package org.apache.kafka.common.errors;
+
+public class BrokerNotAvailableException extends InvalidMetadataException {
```
This is not really an instance of InvalidMetadataException.

ijuma replied:
Is it a `RetriableException` or `ApiException` then?

junrao replied:
ApiException for both.

junrao commented on clients/src/main/java/org/apache/kafka/common/errors/ControllerMovedException.java:
```diff
@@ -0,0 +1,32 @@
+
+package org.apache.kafka.common.errors;
+
+public class ControllerMovedException extends InvalidMetadataException {
```
This is not really an instance of InvalidMetadataException.

ijuma replied:
Is it a `RetriableException` or `ApiException` then?

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/UpdateMetadataRequest.java:
```diff
@@ -0,0 +1,291 @@
+     * Constructor for version 0.
+     */
+    @Deprecated
+    public UpdateMetadataRequest(int controllerId, int controllerEpoch, Set<BrokerEndPoint> aliveBrokers,
```
In version 0, we should only allow passing in a single BrokerEndPoint.

ijuma replied:
Will fix.

ijuma replied:
Are you sure about this? Looking at the Scala code, it does:

Read:
`case 0 => for(i <- 0 until numAliveBrokers) yield new Broker(BrokerEndPoint.readFrom(buffer),SecurityProtocol.PLAINTEXT)`

Write:
``case 0 => aliveBrokers.foreach(_.getBrokerEndPoint(SecurityProtocol.PLAINTEXT).writeTo(buffer))`

junrao replied:
Actually, my comment wasn't right. What you had is right since in v0, we pass in a set of BrokerEndPoint and v1, we pass in a set of Broker.

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -165,70 +203,79 @@ class RequestSendThread(val controllerId: Int,
-        channel.disconnect()
-        error("Controller %d's connection to broker %s was unsuccessful".format(controllerId, broker.toString()), e)
+
+      if (networkClient.isReady(brokerNode, time.milliseconds()))
```
Is the if statement needed since the test is already done in networkClient.blockingReady()?

ijuma replied:
It's done this way to only log the "Controller %d connected" message if the connection is not already ready.

junrao replied:
Ok. We can leave it as it is. Thanks for the explanation.

ijuma committed:
Fix inheritance of `BrokerNotAvailableException` and `ControllerMovedException`

As suggested by Jun.

ijuma committed:
Use `controller-channel` as metrics group prefix.

As suggested by Jun.

ijuma commented:
Pushed a couple of commits addressing review comments.

ijuma committed:
Document `MetadataUpdater`, `ManualMetadataUpdater` and `NetworkClientBlockingOps`

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-2411-remove-usage-of-blocking-channel

* apache/trunk:
  KAFKA-2332; Add quota metrics to old producer and consumer

ijuma commented:
@junrao, merged trunk and added missed scaladoc/javadoc.

`gradlew test` passed locally (as expected since there hasn't been any change that would affect behaviour)

ijuma commented:
@junrao started a Kafka 0.8.2.1 cluster with 3 brokers manually and did a rolling upgrade (with kill -9) to to this branch. Used console producer and console consumer as well as describe-topic and things seemed to work as expected.

junrao commented:
Thanks for the latest patch. Just a minor comment on the names of alive_brokers. Otherwise LGTM.

ijuma committed:
Rename `aliveBrokers` to `liveBrokers`

This makes it consistent with `liveLeaders`.

ijuma commented:
@junrao Renamed to live_brokers as agreed. Tests passed locally.
